changes to cbpi-dev-config, logs and the devcontainer.

prash3r opened the pull request:
cbpi requires empty folders inside the config folder to fully function, and this commit adds them to the cbpi-dev-config.
There was a function `check_for_setup(self)` which should check if a `cbpi setup` is necessary, but it missed a lot of these empty folders. I now added all the files known to me to be checked by the function.

While doing this i saw it also checks for a possible config backup zip `restored_config.zip` and restores it. This was done AFTER checking for config.yaml, it is now done checked and restored bofre every other check.

On unsuccessfull restore the zip file is renamed instead of deleted. So the file isnt lost but the next `cbpi start` will ignore this zip file.

i also added another run command to the vscode config so we can execute the `cbpi setup` from the dev environment easily

and i added the cbpi-dev-config folder to the .gitignore file.

papauorg commented on .gitignore:
```diff
@@ -17,4 +17,5 @@ node_modules
-.coverage
\ No newline at end of file
+.coverage
+.devcontainer/cbpi-dev-config/*
```
The idea behind the cbpi-dev-config folder was, that a ready to go dev environment including common use cases can be provided. This should help to get developers up quickly without needing to setup a entire configuration from scratch when just wanting to get started or review a common scenario. 

When ignoring this folder this is no longer possible. Instead of ignoring it it might be better to provide a valid config that can be used in a better way as it is now (Some sensors/actors? Sample dashboard? Sample hardware that use the sensors? Something else?).

prash3r replied:
Hey, thanks for the feedback,

My intention was to only commit file changes in the config folder if forced with `git add -f` but to ignore changes when using `git add .` . So not every commit spam changes the dev config. But this only works for new files and not for files that are already tracked. Still i think this should stay like that. When we want to commit a working complete config or parts of it we can just do it with `git add -f`. But there wont be any accidentaly commits of additional files in this folder.

What we could do instead: remove everything inside the dev config folder and just put our dev config inside a restored_config.zip file that we force commit. Then when cloning the dev config is extracted but all the new files are not in any commit because they are ignored by the .gitignore. (this would only mark the restored_config.zip as deleted ... dont know what to do with that)

please let me know if i should revert this change.

papauorg replied:
Thanks for the clarification. Sounds good to me. I always like when things like this are obvious to others, maybe we can improve on that: 

- You could add a comment in the gitignore stating that the folder is ignored to keep its clean state and avoid accidentally updating. If updating is intended the dev should execute the step described in the next bullet point
- Create another vscode task "Update development environment config" or similar, that takes the current config and adds all changes and new files under `.devcontainer/cbpi-dev-config` as you described.

I'm not a fan of the zip backup because it hides what was changed and for what reason, because the contents of the zip file are just seen as binary by the git history.

What do you think?

prash3r replied:
Once a file like `config.yaml` is committed (tracked) any changes will be added by `git add .` even if the file is mentioned in the .gitignore file. This is exactly what i misunderstood on my first commit.

I agree that its a bad thing to hide changes inside the zip.

But i like working with vscode tasks. maybe we can commit a dev config in a seperate folder and then use vscode tasks to load this config and overwrite the current dev config or to overwrite the comitted dev config with the current config so it can be commited. Then no files in the actual dev config folder would be commited ever.

papauorg replied:
:+1:

papauorg commented on .vscode/launch.json:
```diff
@@ -6,7 +6,15 @@
+            "request": "launch",
+            "module": "run",
+            "args": ["--config-folder-path=./.devcontainer/cbpi-dev-config", "setup"]
+        },
```
:+1: I like that common tasks are placed in the available tasks in VSCode. That makes it easier to have an overview over what is frequently used in development. Also lets other start more easily. Il ike it!

papauorg commented on cbpi/configFolder.py:
```diff
@@ -115,19 +166,26 @@ def create_config_file(self):
 
     def create_home_folder_structure(configFolder):
-        pathlib.Path(os.path.join(".", 'logs/sensors')).mkdir(parents=True, exist_ok=True)
+        # pathlib.Path(os.path.join(".", 'logs/sensors')).mkdir(parents=True, exist_ok=True)
```
If this line is no longer needed, you should remove it instead of making it a comment. This way it will not confuse future reviewers/developers.

papauorg reviewed (commented):
Is there a reason this PR targets the avollkopf/craftbeerpi4 repository instead of the craftbeerpi/craftbeerpi4 repository. As far as I know @avollkopf could also accept PRs there.

prash3r commented:
with [e61fbdb](https://github.com/avollkopf/craftbeerpi4/pull/62/commits/e61fbdb69cb5290295be937117b07819f18e17c2) i removed the lines that were commented out anyway and reordered the .vscode task profiles so running is the default and not setup hopefully.

prash3r commented on cbpi/configFolder.py:
```diff
@@ -115,19 +165,24 @@ def create_config_file(self):
         print("Config Folder created")
 
     def create_home_folder_structure(configFolder):
-        pathlib.Path(os.path.join(".", 'logs/sensors')).mkdir(parents=True, exist_ok=True)
```
i just noticed that this was intentional. the logs are of course not meant to be inside the config folder but in the folder wherever the user did `cbpi setup`. Im going to change this back.

prash3r replied:
should we really put the logs where `cbpi setup` is called? Then we have to call `cbpi start` from the same folder every time or the log folder is in the wrong place.

i would rather place the log folder in the parent folder of the provided config. This makes more sense especially when we can provide a config folder path somewhere completely different.

For example in the devcontainer setup this would mean the logs folder is inside the .devcontainer folder instead of the repository root.

I will implement the log folder to always be in the parent directory of the provided config folder if there are no objections.

papauorg replied:
How about a separate config variable for the log location similar to the config directory now?

prash3r replied:
i was thinking about this as well but it would break portability of the config. If you provide me with a zipped config for me to test/debug something, I restore it in my docker and then the check setup fails because i dont have the path /user/home/papauorg/cbpi/logs in my docker container. So i dont even get to logging and need to manually alter config files.

its the same when i transfer my own real setup to my docker container to test soemthing.

papauorg replied:
> i was thinking about this as well but it would break portability of the config

I don't think it would break portability of the config because the logs are not part of the config (or at least should not be). 

> If you provide me with a zipped config for me to test/debug something, I restore it in my docker and then the check setup fails because i dont have the path /user/home/papauorg/cbpi/logs in my docker container.

I didn't mean that the path to the log directory should be written anywhere in the config-files. What I meant was, that you can start cbpi with a parameter that specifies where the logs should be written to. I don't really understand why the log directory is relevant in your sample. All you need to run cbpi would be the config. And I'd avoid absolute paths whenever possible.

If you can not start cbpi if the log directory is not present, maybe that's something that should be changed. Either create the log directory if it does not exist or do not log in a file. But not starting the program because the directory does not exist does not make sense to me.

So for example the start command of cbpi for the dev environment (in launch.json) could look something like this:
`"args": ["--config-folder-path=./.devcontainer/cbpi-dev-config", "--log-folder-path=./.devcontainer/logs" "start"]`

I use docker also for running the environment for the actual brewery. I'd not want the logs to be part of the config because I don't want them to clutter my config which I store in git. Logs are pure runtime info and should not be mixed with the config IMHO. With the config parameter I could simply place the log to another volume or something like that and hopefully never look at it :smile:

prash3r replied:
> I'd not want the logs to be part of the config because I don't want them to clutter my config which I store in git

thats why i want to put it in the parent directory

> If you can not start cbpi if the log directory is not present, maybe that's something that should be changed.

problem with the log folder not being present is that then the sensor data cannot be stored and there is no history available and re is no sensor history stored and no charts etc. So especially for someone starting this on a Pi as intended and as documented and not having charts is a bad thing. So the log folder should be created on `cbpi setup` . And when a config is transferred and the log folder isnt present the `cbpi start` informs you that a `cbpi setup` would resolve the issue and terminates.

I want to implement the `--log-folder-path=` option and when it is not provided i will create a log folder in the parent directory of the config folder. This means in the devcontainer the default config will be in the .devcontainer folder and when the `--config-folder-path=` argument is not provided on `cbpi setup` then the log folder as well as the default config folder will both be created at the current directory.

papauorg replied:
> I want to implement the --log-folder-path= option and when it is not provided i will create a log folder in the parent directory of the config folder. This means in the devcontainer the default config will be in the .devcontainer folder and when the --config-folder-path= argument is not provided on cbpi setup then the log folder as well as the default config folder will both be created at the current directory.

Sounds good. You can easily provide a default value for the option so you don't have to care if it's provided or not. Same is done for the config folder path which defaults to "./config" as it was hardcoded before the option was available. See here:

https://github.com/prash3r/craftbeerpi4/blob/4952861a584224311c4072c78ce15abca8913022/cbpi/cli.py#L228

Thanks for your patience with my comments :slightly_smiling_face:

prash3r commented:
i implemented the `--logs-folder-path` cli option. if it is not supplied it will create the `logs` folder right next to the config folder. This applies weather `--config-folder-path` option is used or not. So if neither of the options is used on `cbpi setup` it will create both folders `config` and `logs` in the same current working directory.

and because i was at it i added the cli option `--debug-log-level` which defaults to the old INFO level.

I also made  a change where i set the default log formatter to be as usual but also to be used as default for future log handlers. but if the shell command `journalctl` is found it will remove the timestamp of the streamHandler writing to stdout. This has two effects:

1. since journalctl adds its own timestamps we dont have our downloaded logs begin with 2 identical (but differently formatted) timestamps anymore. 
2. we still have timestamps on stdout when working in the devcontainer.

prash3r commented:
There is one more commit coming.

1. I will rename the `cbpi-dev-config` folder to `cbpi-default-dev-config`
2. and when pressing f5 in vscode i will check via tasks if the folder `cbpi-dev-config` exists and if it doesnt exist i copy the folder `cbpi-default-dev-config` to `cbpi-dev-config` before `cbpi --config-folder-path=./.devcontainer/cbpi-dev-config start` is executed

this way we still have a dev config that is simply run by default in the dev container with only pressing F5 after clone. And any changes we do to the dev config while testing doesnt land in any commit. And if we want the new dev config to be the default we can just delete the `cbpi-default-dev-config` and rename our current config folder `cbpi-dev-config` to `cbpi-default-dev-config`. (overwriting the default with the current dev config can later be done via vscode tasks as well but i wont do it yet)

prash3r commented:
@papauorg i think i have taken all your remarks into account, when there are no further remarks coming i would be done. @avollkopf please let me know if i should squash these commits or if something else should be done here before it can be accepted.

```
git clone -b dev-config-changes https://github.com/prash3r/craftbeerpi4.git prash3rs-dev-config-changes

cd prash3rs-dev-config-changes

code .
```

F5


if anyone could test this, that would be great, please report your findings here.

prash3r commented:
with this last commit we should be able to supply the cbpi-default-dev-config with additional packages that are needed to run the default-dev-config. These additional packages are not installed in the production container.

This is perfect for dev specific packages like my shameless plug for

[cbpi4-SimulatedSensor](https://github.com/prash3r/cbpi4-SimulatedSensor)

which are needed for proper testing in the dev environment but should never be installed in production.

on a later commit we can maybe even commit a complete default-dev-config with dashboard and a recipe and have the recipe basically running immediatly on executing the default dev config with heating towards the target via the simulated sensor.

prash3r commented:
additionally to some minor repairs i figured out how to run the tests locally, modified them to respect the dynamic log folder paths and they pass on my machine now. I would appriceate if you could rerun the workflows.

prash3r commented:
to summerize the changes in this pull request so far:

- moves `cbpi-dev-config` to `cbpi-default-dev-config` and now copies the default on first run via vscode tasks in the dev container.
- `.devcontainer/Dockerfile` installs new `additional-dev-requirements.txt` inside the `.devcontainer/cbpi-default-dev-config/` folder.
- adds the `cbpi-dev-config` to the `gitignore` file so adjusting the dev config locally is not tracked by git
- adds `.vscode/launch.json` run profiles for
  - cbpi plugin creation in debugger
  - cbpi setup in debugger
  - run tests in debugger
- replaces underscores with a dashes in cbpi4 create command
- adds cli option `--logs-folder-path`
- store `configFolderPath` and `logsFolderPath` inside the `ConfigFolder` class and use it wherever it is needed.
- adds function `get_dashboard_path` to `ConfigFolder` class and use it throughout the dashboard controller
- `check_for_setup()` now checks for all the needed files and respects the dynamic logs folder path
- `log_file_controller.py` also respects the dynamic logs folder path
- there is a real cbpi.log file written now with rotational file handler
- when exporting logs from `journalctl` we now use the additional option `--output cat` (not 2 timestamps per line anymore)
- modified tests to supply dynamic logs folder path
- logs generated by the tests are now stored inside the tests folder (and still ignored by git).

please let me know if i should revert some of these changes or squash it into one commit. Further testing/review would be appreciated.
